storage/store.go

```diff
@@ -938,27 +946,33 @@ func (s *Store) NewRangeDescriptor(start, end proto.Key, replicas []proto.Replic
 
 // SplitRange shortens the original range to accommodate the new
 // range. The new range is added to the ranges map and the rangesByKey
-// sorted slice.
+// btree.
 func (s *Store) SplitRange(origRng, newRng *Range) error {
 	if !bytes.Equal(origRng.Desc().EndKey, newRng.Desc().EndKey) ||
 		bytes.Compare(origRng.Desc().StartKey, newRng.Desc().StartKey) >= 0 {
 		return util.Errorf("orig range is not splittable by new range: %+v, %+v", origRng.Desc(), newRng.Desc())
 	}
+
+	s.mu.Lock()
+	defer s.mu.Unlock()
 	// Replace the end key of the original range with the start key of
-	// the new range.
+	// the new range. Reinsert the range since the btree is keyed by range end keys.
+	if s.rangesByKey.Delete(origRng) == nil {
+		return util.Errorf("couldn't find range %s in rangesByKey btree", origRng)
+	}
+
 	copyDesc := *origRng.Desc()
 	copyDesc.EndKey = append([]byte(nil), newRng.Desc().StartKey...)
-	if err := origRng.setDesc(&copyDesc); err != nil {
-		return err
+	origRng.setDescWithoutProcessUpdate(&copyDesc)
```

Comment:
Yes, `setDescLocked` also works. One drawback is that we need to add `processRangeDescriptorUpdateLocked` to the `rangeManager` interface to call it from `setDescLocked`... 

```
func (r *Range) setDescLocked(desc *proto.RangeDescriptor) error {
	atomic.StorePointer(&r.desc, unsafe.Pointer(desc))
	if r.rm == nil {
		// r.rm is null in some tests.
		return nil
	}
	return r.rm.processRangeDescriptorUpdateLocked(r)
}
```